packages/graphql-codegen-cli/src/utils/patterns.ts

```diff
@@ -0,0 +1,308 @@
+      if (typeof doc === 'string') {
+        patterns.push(doc);
+      } else {
+        patterns.push(...Object.keys(doc));
+      }
+    }
+  }
+
+  return patterns;
+};
+
+/**
+ * Parse a list of micromatch patterns from a list of schemas, which should
+ * already have been normalized from their raw config values.
+ */
+const makePatternsFromSchemas = (schemas: Types.Schema[]): string[] => {
+  const patterns: string[] = [];
+
+  for (const s of schemas) {
+    const schema = s as string;
+    if (isGlob(schema) || isValidPath(schema)) {
+      patterns.push(schema);
+    }
+  }
+
+  return patterns;
+};
+
+/**
+ * Given a list of micromatch patterns, sort them into `patterns` (all of them),
+ * `affirmative` (only the affirmative patterns), and `negated` (only the negated patterns)
+ *
+ * @param patterns List of micromatch patterns
+ */
+export const sortPatterns = <P extends string | NegatedPattern>(patterns: P[]): SortedPatterns<P> => ({
+  patterns,
+  affirmative: onlyAffirmativePatterns(patterns) as P[],
+  negated: onlyNegatedPatterns(patterns) as Extract<P, NegatedPattern>[],
+});
+
+/**
+ * A type that "sorts" (or "groups") patterns. For a given list of `patterns`,
+ * this type will include the original list in `patterns`, all of its
+ * "affirmative" (non-negated) patterns in `affirmative`, and all of its
+ * "negated" patterns in `negated`
+ */
+type SortedPatterns<PP extends string | NegatedPattern = string | NegatedPattern> = {
+  /** List of patterns, which could include both negated and affirmative patterns */
+  patterns: PP[];
+  /** List of only the affirmative (non-negated) patterns in `patterns` */
+  affirmative: PP[];
+  /** List of only the negated patterns in `patterns` */
+  negated: Extract<PP, NegatedPattern>[];
+};
+
+/**
+ * The global (top-level) config and each local (output target) config can have
+ * patterns which are separable into "watch" (always takes precedence), "documents",
+ * and "schemas". This type can hold sorted versions of these patterns.
+ */
+type PatternSet = {
+  watch: SortedPatterns;
+  documents: SortedPatterns;
+  schemas: SortedPatterns;
+};
+
+/**
+ * Return true `true` if `relativeCandidatePath` matches at least one pattern
+ * in `patterns` AND is not negated by any negation pattern in `patterns`
+ *
+ * @param relativeCandidatePath A relative path to evaluate against the patterns
+ */
+const pathMatches = (relativeCandidatePath: string, patterns: string[]) => {
+  if (isAbsolute(relativeCandidatePath)) {
+    throw new Error('pathMatches should only be called with relative path');
+  }
+
+  // We use the default micromatch export because it is the only function that
+  // implements the compound matching logic as desired (.isMatch does not remove negations).
+  // This means we don't get a nice boolean result and must check array length instead
+  const match = mm([relativeCandidatePath], patterns, { cwd: process.cwd() }).length === 1;
+
+  return match;
+};
```

Comment:
This is the matcher function we use when we need to match a set of patterns that can include negations. Ideally we could use `micromatch.isMatch`, but as described in #9270, this does not work as expected when negations are included (it treats a negation match as being `true` when it should instead omit it from the match set).

However, the [`micromatch` default export (source here)](https://github.com/micromatch/micromatch/blob/master/index.js#L27-L77) does what we want. It's just a bit awkward because it's not meant to be called as a boolean, so we need to give it a single-item list and then check whether the result has length 1.

Unfortunately, `micromatch` does not implement path normalization, and if patterns are relative, then path must be relative too; `micromatch` will not match an absolute path against relative patterns, nor will it match a relative path against absolute patterns. So, we assume that all patterns are relative paths.  _(Ideally, we should warn the user if any absolute patterns are detected, and I thought I had added code for that, but I just realized it was part of an earlier function since removed. I'll add that warning in a later commit.)_

Note that the thrown errors can only be thrown due to developer error (they check path, not patterns). This is true in general for all thrown errors in this PR; user errors can only result in warnings.